src/persist/src/indexed/encoding.rs

```diff
@@ -26,14 +26,71 @@ use timely::progress::{Antichain, Timestamp};
 use timely::PartialOrder;
 
 use crate::error::Error;
+use crate::gen::persist::proto_batch_part_inline::FormatMetadata as ProtoFormatMetadata;
 use crate::gen::persist::{
     ProtoBatchFormat, ProtoBatchPartInline, ProtoU64Antichain, ProtoU64Description,
 };
 use crate::indexed::columnar::parquet::{decode_trace_parquet, encode_trace_parquet};
-use crate::indexed::columnar::ColumnarRecords;
+use crate::indexed::columnar::{ColumnarRecords, ColumnarRecordsStructuredExt};
 use crate::location::Blob;
 use crate::metrics::ColumnarMetrics;
 
+/// Column format of a batch.
+#[derive(Debug, Copy, Clone)]
+pub enum BatchColumnarFormat {
+    /// Rows are encoded to `ProtoRow` and then a batch is written down as a Parquet with a schema
+    /// of `(k, v, t, d)`, where `k` are the serialized bytes.
+    Row,
+    /// Rows are encoded to `ProtoRow` and a columnar struct. The batch is written down as Parquet
+    /// with a schema of `(k, k_c, v, v_c, t, d)`, where `k` are the serialized bytes and `k_c` is
+    /// nested columnar data.
+    Both,
+}
+
+impl BatchColumnarFormat {
+    /// Returns a default value for [`BatchColumnarFormat`].
+    pub const fn default() -> Self {
+        // IMPORTANT: Default to only writing Row data and not our newer structured format.
+        BatchColumnarFormat::Row
+    }
+
+    /// Returns a [`BatchColumnarFormat`] for a given `&str`, falling back to a default value if
+    /// provided `&str` is invalid.
+    pub fn from_str(s: &str) -> Self {
+        match s {
+            "row" => BatchColumnarFormat::Row,
+            "both" => BatchColumnarFormat::Both,
+            x => {
+                let default = BatchColumnarFormat::default();
+                tracing::error!("Invalid batch columnar type: {x}, falling back to {default}");
```

Comment:
ditto should do something that fails CI here

Reply:
Started using `soft_panic_or_log!` macro which does exactly that